openfisca_core/indexed_enums/_utils.py

```diff
@@ -0,0 +1,216 @@
+import numpy
+
+from . import types as t
+from ._errors import EnumMemberNotFoundError
+
+
+def _enum_to_index(enum_class: type[t.Enum], value: t.ObjArray) -> t.IndexArray:
+    """Transform an array of enum members into an index array.
+
+    Args:
+        enum_class: The enum class to encode the enum members array.
+        value: The enum members array to encode.
+
+    Returns:
+        The index array.
+
+    Raises:
+        EnumMemberNotFoundError: If one value is not in the enum class.
+
+    Examples:
+        >>> import numpy
+
+        >>> from openfisca_core import indexed_enums as enum
+
+        >>> class Road(enum.Enum):
+        ...     STREET = (
+        ...         "A public road that connects two points, but also has "
+        ...         "buildings on both sides of it; these typically run "
+        ...         "perpendicular to avenues."
+        ...     )
+        ...     AVENUE = (
+        ...         "A public way that also has buildings and/or trees on both "
+        ...         "sides; these run perpendicular to streets and are "
+        ...         "traditionally wider."
+        ...     )
+
+        >>> class Rogue(enum.Enum):
+        ...     BOULEVARD = "More like a shady impasse, to be honest."
+
+        # >>> _enum_to_index(Road, numpy.array(Road.AVENUE))
+        # array([1], dtype=uint8)
+        #
+        # >>> _enum_to_index(Road, numpy.array([Road.AVENUE]))
+        # array([1], dtype=uint8)
+        #
+        # >>> value = numpy.array([Road.STREET, Road.AVENUE, Road.STREET])
+        # >>> _enum_to_index(Road, value)
+        # array([0, 1, 0], dtype=uint8)
+
+        >>> value = numpy.array([Road.AVENUE, Road.AVENUE, Rogue.BOULEVARD])
+        >>> _enum_to_index(Road, value)
+        Traceback (most recent call last):
+        EnumMemberNotFoundError: Member BOULEVARD not found in enum 'Road'...
+
+    """
+    # Create a mask to determine which values are in the enum class.
```

Comment:
@benjello This works as expected, and I'm currently fixing bad tests in France. This is orders of magnitue faster than before. However, it is verbose. If you see how to make this simpler, do not hesitate.

Reply:
Could you elaborate on why this is faster ?

Reply:
I'm afraid of the loss of performance on large data.

Reply:
What I'm saying if this should be faster than previously, as we're relying purely on numpy.

Reply:
(We can, if the _if_ creates a perf hog, do as before: silently default to 0 on errors, or well give a -1. In any case, I understood defaulting to 0 as a bug. If it is not, I will adapt this, otherwise this would be a breaking change).

Reply:
Oh I see, `numpy.isin` is terribly slow. I'll fix that.

Reply:
I've been skimming this further but I'm still ~50% overhead of master in encode. Tue problem space is weird because I find no way of avoiding data copy (instead of reusing a buffer or memory view).

Reply:
Ok, now it is x5-x10 faster than before.